x/coin.go

```diff
@@ -86,11 +90,17 @@ func (c Coin) Multiply(times int64) Coin {
 		frac = frac % FracUnit
 	}
 
-	return Coin{
+	wantPositive := times > 0 == (c.Whole >= 0 && c.Fractional >= 0)
+	if whole >= 0 && frac >= 0 && !wantPositive {
```

Comment:
I would do the same check as above...

```golang
if res.IsNonNegative() != wantPositive {
  return Coin{}, errors.ErrOverflow
}
```

Reply:
However, I think this may be the pseudo-code you'd prefer rather than this sign check...

```
if whole / times != c.Whole {
  return Coin{}, errors.ErrOverflow.New("whole component")
}
if frac / times != c.Fractional {
  return Coin{}, errors.ErrOverflow.New("fractional component")
}
```

Ref: https://stackoverflow.com/a/1815371

Reply:
This code is gone. :see_no_evil: